Disabling agent account doesn’t terminate active connections - Fix

SudhanshuBawane opened the pull request:
closed [2608](https://github.com/sensu/sensu-enterprise-go/issues/2608)

## Description

To fix the behavior of the agent not getting disconnected while user is disabled.


## Fixed

The agent is now getting disconnected as soon as the user is disabled unlike before.


## Added

- The new watcher configuration for user has been added.
- The session code has been modified to consume the new user configuration into the bus.


## Change verification

To verify the change as normal run backend and agent, as soon as both are running and connected, Disable the user via sensuctl command. In backend you will see a shutting down a agent session prompt and in agent will fail to do a handshake with bad credential prompt.

SudhanshuBawane committed:
initial

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
inital

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
change

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

fguimond committed:
Sample code to use the wizard bus to tie the user watcher with the
session shutdown

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
WIP

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
unmarshal issue

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
testing

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
testing

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
working solution

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
with test cases

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
with test cases

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
final

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
testing

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
testing

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
final

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

SudhanshuBawane committed:
final

Signed-off-by: SudhanshuBawane <sudhanshu.bawane.ctr@sumologic.com>

rakibhossainctr committed:
merge conflict resolved

Signed-off-by: rakibhossainctr <rakib.hossain.ctr@sumologic.com>

rakibhossainctr committed:
refactor: changelog shifting to unreleased

Signed-off-by: rakibhossainctr <rakib.hossain.ctr@sumologic.com>

rakibhossainctr committed:
removed panic statement

Signed-off-by: rakibhossainctr <rakib.hossain.ctr@sumologic.com>

rakibhossainctr committed:
refactor: removed comments from test case

Signed-off-by: rakibhossainctr <rakib.hossain.ctr@sumologic.com>

rakibhossainctr commented:
ready to merge

rakibhossainctr committed:
Merge branch 'develop/6' into sudhanshu#2608-fix

chavakula commented on backend/agentd/agentd.go:
```diff
@@ -308,6 +320,24 @@ func (a *Agentd) handleEvent(event store.WatchEventEntityConfig) error {
 
+// adding the UserConfig updates to the etcd bus for the watcher to consume
+func (a *Agentd) handleUserEvent(event *store.WatchEventUserConfig) error {
+	if event.User == nil {
```
Check if event is null here, if event is null it will panic

chavakula commented on backend/agentd/session.go:
```diff
@@ -321,6 +347,45 @@ func (s *Session) sender() {
+			}
+
+			if watchEvent.User == nil {
+				logger.Error("session received nil user in watch event")
```
You should continue or return after detecting nil user

chavakula commented on backend/agentd/session.go:
```diff
@@ -321,6 +347,45 @@ func (s *Session) sender() {
+			if watchEvent.User == nil {
+				logger.Error("session received nil user in watch event")
+			}
+			lagger := logger.WithFields(logrus.Fields{
```
lagger => logger

chavakula commented on backend/agentd/session.go:
```diff
@@ -321,6 +347,45 @@ func (s *Session) sender() {
+			})
+			lagger.Debug("user update received")
+
+			bytes, err := s.marshal(watchEvent.User)
```
bytes is a standard library package name.  avoid using them

chavakula commented on backend/agentd/watcher.go:
```diff
@@ -70,3 +69,44 @@ func GetEntityConfigWatcher(ctx context.Context, client *clientv3.Client) <-chan
+			// unmarshal the user config
+			var userConfig corev2.User
+			if err := proto.Unmarshal(response.Object, &userConfig); err != nil {
+				continue
```
Log error message here

rakibhossainctr committed:
fix: address PR review feedback for user config watcher

- agentd.go: add nil check for event before accessing event.User in
  handleUserEvent to prevent panic on nil pointer dereference
- session.go: add continue after nil user check to prevent panic,
  rename lagger to logger (typo fix), rename bytes to userData to
  avoid shadowing standard library package, add continue after
  marshal error
- watcher.go: add error logging when proto.Unmarshal fails for user
  config

Signed-off-by: rakibhossainctr <rakib.hossain.ctr@sumologic.com>
